### PR TITLE
Add filters to issue list

### PR DESCRIPTION
**Pull request by ari7946**
Implements Filters by level, status, and by partially typing a project's name. Cypress tests included.

### CODE REVIEW

**Review by mentor-tara — APPROVED** [review]
Go go go :smile:[](prolog-34)

File: cypress/integration/issue-list.spec.ts [hunk 1]
```diff
@@ -95,5 +100,65 @@ describe("Issue List", () => {
           });
         });
     });
+
+    it("renders issues with resolved status only", () => {
+      cy.dataCy("filter-by-status").click();
+      cy.wait(2000);
```
Comment (jkettmann) [inline cypress/integration/issue-list.spec.ts:106]: Did you try the tests without waiting manually? Cypress should wait by itself so maybe these aren't necessary.

File: cypress/support/commands.js [hunk 2]
```diff
@@ -23,3 +23,21 @@
 //
 // -- This will overwrite an existing command --
 // Cypress.Commands.overwrite('visit', (originalFn, url, options) => { ... })
+Cypress.Commands.add("validateIssues", (mockedIssues) => {
```
Comment (jkettmann) [inline cypress/support/commands.js:26]: Nice idea to add this as a command 👍

File: features/issues/api/use-issues.tsx [hunk 3]
```diff
@@ -3,18 +3,28 @@ import { useQuery, useQueryClient } from "react-query";
...
 import type { Issue } from "../types/issue.types";
+import { useFilters } from "@features/issues";
+import { IssueFilters } from "@features/issues";
 
-async function getIssues(page: number) {
+async function getIssues(page: number, filters: IssueFilters) {
   const { data } = await axios.get(
-    `https://prolog-api.profy.dev/issue?page=${page}`
+    `https://prolog-api.profy.dev/issue?page=${page}`,
+    { params: filters }
```
Comment (jkettmann) [inline features/issues/api/use-issues.tsx:12]: 👍

File: features/issues/api/use-issues.tsx [hunk 4]
```diff
@@ -3,18 +3,28 @@ import { useQuery, useQueryClient } from "react-query";
...
   return data;
 }
 
 export function useIssues(page: number) {
+  const { filters } = useFilters();
+
+  useEffect(
+    () => console.log("update filters!"),
+    [filters.level, filters.project, filters.status]
+  );
```
Comment (jkettmann) [inline features/issues/api/use-issues.tsx:23]: This seems to be a left-over

File: features/issues/components/IssueList/Issue-list.tsx [hunk 5]
```diff
@@ -94,49 +96,70 @@ export function IssueList() {
...
+          </thead>
+          <tbody>
+            {(items || [])
+              // .filter(
+              //   (issue) =>
+              //     inputValue === "" ||
+              //     projectIdToName[issue.projectId]
+              //       .toLowerCase()
+              //       .includes(inputValue.toLowerCase())
+              // )
```
Comment (jkettmann) [inline features/issues/components/IssueList/Issue-list.tsx:131]: I'd simply remove code that's commented out. Looks better and avoids future confusion

File: features/issues/components/filters/filters.tsx [hunk 6]
```diff
@@ -0,0 +1,101 @@
...
+      handleProjectName(undefined);
+      return;
+    }
+    const name = projectNames?.find(
+      (name) => inputValue.length > 2 && name.includes(inputValue)
+    );
+
+    setProjectName((prevName) => name || prevName);
+    handleProjectName(projectName);
+  }, [inputValue]);
```
Comment (jkettmann) [inline features/issues/components/filters/filters.tsx:50]: I wonder if you could simply add this logic to the `handleChange` function. It's mostly not necessary to run a `useEffect` upon a state change.

File: features/issues/components/filters/filters.tsx [hunk 7]
```diff
@@ -0,0 +1,101 @@
...
+  return (
+    <Container>
+      <Select
+        placeholder="Status"
+        defaultValue="Status"
+        width="8rem"
+        data-cy="filter-by-status"
+      >
+        <Option value={undefined} handleCallback={handleStatus}>
+          --None--
```
Comment (jkettmann) [inline features/issues/components/filters/filters.tsx:61]: Nice

File: features/issues/components/filters/filters.tsx [hunk 8]
```diff
@@ -0,0 +1,101 @@
...
+  display: flex;
+  gap: 3rem;
+  justify-content: flex-end;
+`;
+
+export function Filters() {
+  const { handleFilters } = useFilters();
+  const { data: projects } = useProjects();
+  const [inputValue, setInputValue] = useState("");
+  const [projectName, setProjectName] = useState<string | undefined>(undefined);
```
Comment (jkettmann) [inline features/issues/components/filters/filters.tsx:18]: I'm curious why you have the `projectName` state? It doesn't seem to be used anywhere

**Comment by jkettmann** [comment]
Great job building this feature. I'm curious about two things:

1. Are the filters added to the URL as query params? That's the last acceptance criterium of the task. This also simplifies the update logic of the filters as you can simply push the new query params to the URL from the filters component and the issue list component is updated automatically.
2. I think each change of the project name input triggers a request, is that right? You can use debouncing to limit the number of requests. Have a look at lodash.debounce for example

File: features/issues/components/filters/filters.tsx (hunk 8 above)
Comment (ari7946) [reply features/issues/components/filters/filters.tsx:18]: you're right. It is not necessary and has been removed.

**Comment by ari7946** [comment]
Thanks.
1:  They weren't added before but now the filters have been added to the URL as query params.
2: It did cross my mind to use debounce but it doesn't exactly query on every input change so I thought it would be unnecessary. It uses this logic
```
    setInputValue(input);

    if (inputValue?.length < 2) {
      handleProjectName(undefined);
    }

    const name = projectNames?.find(
      (name) =>
        inputValue?.length > 2 &&
        name?.toLowerCase().includes(inputValue.toLowerCase())
    );

    if (name) {
      handleProjectName(name);
    }
    ```
    If a user types "back" it will trigger a request with the project name "backend", then if a user continues typing "backend", there would be no additional requests. A request will be triggered again only when a user deletes most characters(which would then make a request for all projects by removing the filter).

**Comment by jkettmann** [comment]
Ah now I get it. Smart approach. Kind of an autocomplete, right? I didn't try it out yet, but does it also fill in the project name in the input? And what if there are two projects with the same subset of letters. E.g. "backend-node" and "backend-python". The current implementation would get whatever comes first in the array. Am I assuming right, that the user would simply continue typing and when they're at "backend-p" it would find the python project?

I think this is a good idea. You just have to be careful not to introduce bugs or unwanted/unexpected behavior. The endpoint also offers a search function iirc. So if you search for "back" it should return issues for both of the "backend" projects in the example above.

**Comment by ari7946** [comment]
- **Kind of an autocomplete, right?** yes
-**...does it also fill in the project name in the input?** It currently doesn't autocomplete the name in the input; it only autocompletes the project query param. It does, however, pre-fill the input when a user selects "View issues" from the projects page.
- **And what if there are two projects with the same subset of letters. E.g. "backend-node" and "backend-python". The current implementation would get whatever comes first in the array. Am I assuming right, that the user would simply continue typing and when they're at "backend-p" it would find the python project?** Yes, you're right that it would return the first project in the array and also that as soon as they finish typing "backend-p", it would return the python project if it wasn't already the first in the array with "backend". I think I this could be a problem, perhaps instead of a `find` method, it should be a `filter`, and only autocomplete when there is one item in the array.

As for the second paragraph. Yeah, I think I already introduced bugs 😄. For example the test that tests for the page number to persist after a reload stopped working. Hmmm that search function iirc sounds nice. I didn't realize you can search by partial names and this is why implemented the autocomplete in the first place lol.

**Comment by ari7946** [comment]
Edit^^ above. I just edited the above comment with more thoughts about the input field.

**Pull request merged**